api/src/main/java/org/infinispan/api/common/CacheOptions.java

```diff
@@ -45,36 +52,41 @@ public Duration rawTimeout() {
       }
 
       @Override
-      public Optional<EnumSet<?>> flags() {
+      public Optional<Flags<?, ?>> flags() {
          return Optional.ofNullable(flags);
       }
 
-      public EnumSet<?> rawFlags() {
+      public Flags<?, ?> rawFlags() {
          return flags;
       }
    }
 
    class Builder {
       protected Duration timeout;
-      protected EnumSet<?> flags;
+      protected Flags<?, ?> flags;
 
       public Builder timeout(Duration timeout) {
          this.timeout = timeout;
          return this;
       }
 
       public Builder flags(Flag flag) {
-         flags = flag.apply(flags);
+         flags = flag.add(flags);
          return this;
       }
 
       public Builder flags(Flag... flags) {
          for (Flag flag : flags) {
-            this.flags = flag.apply(this.flags);
+            this.flags = flag.add(this.flags);
          }
          return this;
       }
```

Comment:
I would remove these methods and just keep one with `Flags` interface. 
I think we need to have 2 different impls (one for remote and one for embedded) and if you do the mistake of `add(EmbeddedFlag).add(RemoteFlag)`, you probably will have a `RuntimeException` without any compile warning.

Reply:
I have removed the `flags(Flag flag)` and the `flags(Flag... flags)` method, leaving only the `flags(Flags flags)`

Reply:
We will definitely have two implementations, but I don't know how we can prevent passing the wrong `Flags` at compile time...

Reply:
> > I can pack flags immediately into an int.
> 
> I would love to have that but it also opens the possibility to users adding their own implementation.

The only way in which we can prevent other implementations is to use JDK 17's sealed classes, but that means that the `EmbeddedFlags` and the `RemoteFlags` must be included in the API, and you could still use Embedded flags for a remote call and viceversa, which would only be detected at runtime

Reply:
the only thing I can think of is just to have a single `Flags` class with 2 values inside: one for embedded and another for remote.
We could use a single `Flag` enum and each enum's value would have the bit in the embedded or remote (or both) :shrug: